src/core/server/http/router/request.ts

```diff
@@ -138,12 +154,19 @@ export class KibanaRequest<
       enumerable: false,
     });
 
-    this.route = deepFreeze(this.getRouteInfo());
+    this.route = deepFreeze(this.getRouteInfo(request));
     this.socket = new KibanaSocket(request.raw.req.socket);
+    this.events = this.getEvents(request);
+  }
+
+  private getEvents(request: Request): KibanaRequestEvents {
+    const end$ = fromEvent<void>(request.raw.req, 'end').pipe(first());
```

Comment:
I added `end$` to complete `aborted$` observable to prevent memory leaks. I use the `end` event here to make the tests green https://github.com/elastic/kibana/pull/55061/files#diff-7ffe36ae4fa33e0fddf01a3182306eedR47 
From my understanding [end](https://nodejs.org/api/stream.html#stream_event_end) cannot be used here, as it's emitted when a request payload is fully consumed. It seems that I should use [close](https://nodejs.org/api/http.html#http_event_close_2) instead, but switching to `close` fails the tests. I have a question: what is the correct way to understand that request won't emit any new events and I can complete `aborted$`?

Reply:
@elastic/kibana-platform @watson any thoughts?

Reply:
I'm not that familiar with rxjs, but I'll do my best to answer.

You stepped right into the most complicated part of Node.js 😅 Streams can "end" in many ways unfortunately, it differs with the type of stream, and in some cases also with the version of Node.js, and it's often hard to implement correctly without memory leaks - as you also found out.

A normal request/response event-lifecycle looks something like this:

![image](https://user-images.githubusercontent.com/10602/72618336-2d0e7f80-393b-11ea-9b0e-eb9ad5014732.png)

In case the underlying socket is destroyed the order of events are like so:

![image](https://user-images.githubusercontent.com/10602/72618391-40b9e600-393b-11ea-8d0d-0a2652acf468.png)

_Note: If I want to be sure to be notified when a stream ends I always use the [end-of-stream](https://www.npmjs.com/package/end-of-stream) module._

Pipe only cares about the `end` event, so if you pipe `source` into `target`, but `source` then emits a `close` event, `target` will not know about it and the pipe will still be open:

![image](https://user-images.githubusercontent.com/10602/72619471-a4451300-393d-11ea-9ab1-7b189bb8806f.png)

Similar to the end-of-stream module, there's a [pump](https://www.npmjs.com/package/pump) module which takes care of this edge-case, so that the target stream is correctly closed if the source stream closes. **I think that's the issue you're seeing**

> Note: I don't set `connection: keep-alive/close` explicitly in tests, I assume the code should work for either case.

Correct, the events on the request and response objects shouldn't be affected by the `Connection` header.

Reply:
Great explanation, really appreciate it!
I surprised to see `Server: Socket: closed` in both your examples. From my understanding server should keep the socket open since HTTP/1.1 uses `connection: keep-alive` by default. What I missed here? 

https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html
> A significant difference between HTTP/1.1 and earlier versions of HTTP is that persistent connections are the default behavior of any HTTP connection. That is, unless otherwise indicated, the client SHOULD assume that the server will maintain a persistent connection, even after error responses from the server.

What I missed here?

Reply:
In my example, I used the default global HTTP Agent in Node.js which doesn't keep connections alive. If I set it to keep-alive, it looks like this:

![image](https://user-images.githubusercontent.com/10602/72626929-e5dcba80-394b-11ea-89c3-1c70697394d5.png)

Reply:
@watson is this built-in nodejs logger?

Reply:
Nope, I'm using my own event debugging module: https://www.npmjs.com/package/event-debug

Reply:
Then `end` usage in the code looks correct:
- server handles an incoming request and responds. `end` emitted --> `aborted$` completes --> `end$` completes
- an incoming request aborted / socket destroyed. `aborted` emitted --> `aborted$` completes --> `end$` unsubscribes from `end` event (there are no other subscriptions).

Therefore memory a memory leak should not be possible.
Additionally, I performed a loading test locally and everything looks good. [load test](https://github.com/restrry/test-rx-nodejs-events-memory-leak)
Using additional package https://github.com/mafintosh/end-of-stream looks a bit overkill for such a small use-case. 
@elastic/kibana-platform are you okay with the current implementation or we should consider exposing the EventEmitter interface instead of Observables? Then we needn't to think about cleaning up after manually created Observable.